File: Day #520/solution.py
# Day 520: Tree depth = max nesting level of '(' in the representation. O(n) time, O(1) space.
def tree_depth(s):
    depth = best = 0
    for c in s:
        if c == '(':
            depth += 1
            best = max(best, depth)
        elif c == ')':
            depth -= 1
    return best


if __name__ == "__main__":
    print(tree_depth("((((00)0)0)0)"))  # 4

File: Day #520/solution.rs
// Tree depth = max nesting level of '(' in the representation. O(n) time, O(1) space.
fn tree_depth(s: &str) -> i32 {
    let (mut depth, mut best) = (0, 0);
    for c in s.chars() {
        match c {
            '(' => {
                depth += 1;
                if depth > best {
                    best = depth;
                }
            }
            ')' => depth -= 1,
            _ => {}
        }
    }
    best
}

fn main() {
    println!("{}", tree_depth("((((00)0)0)0)")); // 4
}
